# Segmenting and Clustering Neighborhoods in Toronto

# PART 1

## Download and Explore Dataset

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. We will scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format.

Wikipedia page = https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

## Load needed libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis


import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


### Download the data

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df = pd.read_html(url)
df

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                           Neighborhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

Let's check how many tables are on the page.

In [3]:
len(df)

3

There are 3 tables on the page. We select the first one.

In [4]:
df[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [5]:
toronto_df = df[0]
toronto_df.shape

(180, 3)

A dataframe consist of three columns: PostalCode, Borough, and Neighborhood. It has 180 rows.

### Data Cleaning

In [6]:
toronto_df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [7]:
toronto_df['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

As we see above there are 77 cells which doesn't have an assigned borough.

We only process the cells that have an assigned borough. We ignore cells with a borough that is Not assigned. 

In [8]:
toronto_df=toronto_df.loc[toronto_df['Borough']!='Not assigned']
toronto_df.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [9]:
toronto_df['Neighborhood'] = toronto_df['Neighborhood'].replace('Not assigned',toronto_df['Borough'])

/Users/iwonaczech/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [10]:
toronto_df = toronto_df.reindex()
toronto_df.shape

(103, 3)

After conducted operations there are 103 rows left in a dataframe.

### END OF PART 1

#  

# PART 2

## Adding the latitude and the longitude coordinates of each neighborhood

In [11]:
#load the additional data

geo_df = pd.read_csv('/Users/iwonaczech/Downloads/Geospatial_Coordinates.csv')

In [12]:
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Merge 2 dataframes geo_df and toronto_df, by a common column 'Postal Code'

In [13]:
toronto_DF = pd.merge(toronto_df, geo_df, on='Postal Code')
toronto_DF

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [14]:
toronto_DF.shape

(103, 5)

### END OF PART 2

#  

# PART 3

## Explore and cluster the neighborhoods in Toronto

In [15]:
toronto_DF['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
East Toronto         5
York                 5
Mississauga          1
Name: Borough, dtype: int64

In [16]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_DF['Borough'].unique()),
        toronto_DF.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


### Use geopy library to get the latitude and longitude values of Toronto


In [17]:
address = 'Toronto, M'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Create a map of Toronto with neighborhoods superimposed on top

In [18]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_DF['Latitude'], toronto_DF['Longitude'], toronto_DF['Borough'], toronto_DF['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's focus on a subset of all the Borough that contain the word 'Toronto'.

In [19]:
toronto_DF = toronto_DF[toronto_DF['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_DF

Postal Code           Borough  \
0          M5A  Downtown Toronto   
1          M7A  Downtown Toronto   
2          M5B  Downtown Toronto   
3          M5C  Downtown Toronto   
4          M4E      East Toronto   
5          M5E  Downtown Toronto   
6          M5G  Downtown Toronto   
7          M6G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M6H      West Toronto   
10         M5J  Downtown Toronto   
11         M6J      West Toronto   
12         M4K      East Toronto   
13         M5K  Downtown Toronto   
14         M6K      West Toronto   
15         M4L      East Toronto   
16         M5L  Downtown Toronto   
17         M4M      East Toronto   
18         M4N   Central Toronto   
19         M5N   Central Toronto   
20         M4P   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
23         M4R   Central Toronto   
24         M5R   Central Toronto   
25         M6R      West Toronto   
26         M4S   Central Toronto   
27         M5S  Downtown Toronto   
28         M6S      West Toronto   
29         M4T   Central Toronto   
30         M5T  Downtown Toronto   
31         M4V   Central Toronto   
32         M5V  Downtown Toronto   
33         M4W  Downtown Toronto   
34         M5W  Downtown Toronto   
35         M4X  Downtown Toronto   
36         M5X  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                            Garden District, Ryerson  43.657162 -79.378937  
3                                      St. James Town  43.651494 -79.375418  
4                                         The Beaches  43.676357 -79.293031  
5                                         Berczy Park  43.644771 -79.373306  
6                                  Central Bay Street  43.657952 -79.387383  
7                                            Christie  43.669542 -79.422564  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9                        Dufferin, Dovercourt Village  43.669005 -79.442259  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                  North Toronto West, Lawrence Park  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                        Moore Park, Summerhill East  43.689574 -79.383160  
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
31  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -

There are 39 Neighborhoods that it's Borough name contains word Toronto.

In [20]:
toronto_DF.shape

(39, 5)

### Let's create a map of Toronto with Boroughs that cointain the word Toronto


In [21]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_DF['Latitude'], toronto_DF['Longitude'], toronto_DF['Borough'], toronto_DF['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Let's utilize the Foursquare API to explore the neighborhoods and segment them

We need to define Foursquare Credentials and Version.

In [22]:
CLIENT_ID = 'XXX' #Foursquare ID
CLIENT_SECRET = 'XXX' # Foursquare Secret
VERSION = '20200605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: J0X1ZITI2QFGRFSFUHQTYVV1FN31IWDIFW0WZMNRSX3EWNOA
CLIENT_SECRET:BFJEJC0MABHCQPQCBXRF5ERXL5YGSGOQMSQMUPSP4A4SAU3D


In [23]:
# type your answer here
radius=500
LIMIT=100

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,limit=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [25]:
toronto_venues = getNearbyVenues(names=toronto_DF['Neighborhood'],
                                   latitudes=toronto_DF['Latitude'],
                                   longitudes=toronto_DF['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [26]:
print(toronto_venues.shape)
toronto_venues

(1618, 7)


Neighborhood  \
0                             Regent Park, Harbourfront   
1                             Regent Park, Harbourfront   
2                             Regent Park, Harbourfront   
3                             Regent Park, Harbourfront   
4                             Regent Park, Harbourfront   
...                                                 ...   
1613  Business reply mail Processing Centre, South C...   
1614  Business reply mail Processing Centre, South C...   
1615  Business reply mail Processing Centre, South C...   
1616  Business reply mail Processing Centre, South C...   
1617  Business reply mail Processing Centre, South C...   

      Neighborhood Latitude  Neighborhood Longitude                    Venue  \
0                 43.654260              -79.360636         Roselle Desserts   
1                 43.654260              -79.360636            Tandem Coffee   
2                 43.654260              -79.360636       Morning Glory Cafe   
3                 43.654260              -79.360636   Cooper Koo Family YMCA   
4                 43.654260              -79.360636      Body Blitz Spa East   
...                     ...                     ...                      ...   
1613              43.662744              -79.321558     TTC Russell Division   
1614              43.662744              -79.321558  Jonathan Ashbridge Park   
1615              43.662744              -79.321558       Toronto Yoga Mamas   
1616              43.662744              -79.321558         Olliffe On Queen   
1617              43.662744              -79.321558              ONE Academy   

      Venue Latitude  Venue Longitude        Venue Category  
0          43.653447       -79.362017                Bakery  
1          43.653559       -79.361809           Coffee Shop  
2          43.653947       -79.361149        Breakfast Spot  
3          43.653249       -79.358008   Distribution Center  
4          43.654735       -79.359874                   Spa  
...              ...              ...                   ...  
1613       43.664908       -79.322560    Light Rail Station  
1614       43.664702       -79.319898                  Park  
1615       43.664824       -79.324335           Yoga Studio  
1616       43.664503       -79.324768               Butcher  
1617       43.662253       -79.326911  Gym / Fitness Center  

[1618 rows x 7 columns]

### Let's check which Neighborhood has the bigges number of venues

In [27]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     18   
Central Bay Street                                                     65   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             31   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         21   
Kensington Market, Chinatown, Grange Park                              60   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               45   
Moore Park, Summerhill East                                             2   
North Toronto West, Lawrence Park                                      19   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               95   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     38   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         94   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      18   
Central Bay Street                                                      65   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              31   
Davisville North                    

In [28]:
tv = toronto_venues.groupby('Neighborhood').count()
tv['Venue'].sort_values(ascending=False)

Neighborhood
Commerce Court, Victoria Hotel                                                                                100
First Canadian Place, Underground city                                                                        100
Garden District, Ryerson                                                                                      100
Harbourfront East, Union Station, Toronto Islands                                                             100
Toronto Dominion Centre, Design Exchange                                                                      100
Richmond, Adelaide, King                                                                                       95
Stn A PO Boxes                                                                                                 94
Church and Wellesley                                                                                           78
St. James Town                                                             

Neighborhood's with top number of venue's which is 100 are:

Commerce Court, Victoria Hotel                                                                               
First Canadian Place, Underground city                                                                       
Garden District, Ryerson                                                                                      
Harbourfront East, Union Station, Toronto Islands                                                             
Toronto Dominion Centre, Design Exchange 

### Let's examine unique venues categories 

In [29]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


In [30]:
toronto_venues.groupby('Venue Category').count()

Neighborhood  Neighborhood Latitude  \
Venue Category                                            
Afghan Restaurant              1                      1   
Airport                        1                      1   
Airport Food Court             1                      1   
Airport Gate                   1                      1   
Airport Lounge                 2                      2   
...                          ...                    ...   
Wine Bar                       9                      9   
Wine Shop                      3                      3   
Wings Joint                    2                      2   
Women's Store                  1                      1   
Yoga Studio                   13                     13   

                    Neighborhood Longitude  Venue  Venue Latitude  \
Venue Category                                                      
Afghan Restaurant                        1      1               1   
Airport                                  1      1               1   
Airport Food Court                       1      1               1   
Airport Gate                             1      1               1   
Airport Lounge                           2      2               2   
...                                    ...    ...             ...   
Wine Bar                                 9      9               9   
Wine Shop                                3      3               3   
Wings Joint                              2      2               2   
Women's Store                            1      1               1   
Yoga Studio                             13     13              13   

                    Venue Longitude  
Venue Category                       
Afghan Restaurant                 1  
Airport                           1  
Airport Food Court                1  
Airport Gate                      1  
Airport Lounge                    2  
...                             ...  
Wine Bar                          9  
Wine Shop                         3  
Wings Joint                       2  
Women's Store                     1  
Yoga Studio                      13  

[237 rows x 6 columns]

### Venues sorted by a number of its occurrence

In [31]:
toronto_venues['Venue Category'].value_counts()
#toronto_venues.groupby('Venue Category').count()['Venue'].sort_values(ascending=False) # this code does the same as above

Coffee Shop                  143
Café                          92
Restaurant                    54
Italian Restaurant            41
Hotel                         35
                            ... 
College Auditorium             1
Cajun / Creole Restaurant      1
Harbor / Marina                1
Boat or Ferry                  1
Market                         1
Name: Venue Category, Length: 237, dtype: int64

The most popular venue among all of the neighborhoods is Coffee Shop, right behind him Cafe.

What's is the difference between this two?

Generally in a coffee shop, coffee is the main focus. You will find different brewing methods, traditional espresso beverages, and most also have a selection of tea offerings. Officially, a cafe can also be referred to as a restaurant. In cafes, the main focus is on food rather than coffee, though most cafes will offer coffee pairings on their menus.

## Analyze Each Neighborhood

In [32]:
# Apply one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
#toronto_onehot

# Add neighborhood column back to a dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
#toronto_onehot

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]


toronto_onehot.head()
toronto_onehot.columns

Index(['Neighborhood', 'Afghan Restaurant', 'Airport', 'Airport Food Court',
       'Airport Gate', 'Airport Lounge', 'Airport Service', 'Airport Terminal',
       'American Restaurant', 'Antique Shop',
       ...
       'Trail', 'Train Station', 'Vegetarian / Vegan Restaurant',
       'Video Game Store', 'Vietnamese Restaurant', 'Wine Bar', 'Wine Shop',
       'Wings Joint', 'Women's Store', 'Yoga Studio'],
      dtype='object', length=237)

In [33]:
toronto_onehot.shape

(1618, 237)

There are 1618 different venues (rows) which belongs to one of the 237th unique venues category (columns).

### Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [34]:
# need to rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighborhood'}, inplace = True)

toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.012821   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.055556            0.055556      0.055556        0.111111   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.0000

In [35]:
toronto_grouped.shape # Let's check if size is OK

(39, 237)

## Let's print each neighborhoods the top 5 most common venues


In [36]:
top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("*****"+hood+"*****")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(top_venues))
    print('\n')

*****Berczy Park*****
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2  Seafood Restaurant  0.04
3            Beer Bar  0.04
4          Restaurant  0.04


*****Brockton, Parkdale Village, Exhibition Place*****
               venue  freq
0               Café  0.13
1     Breakfast Spot  0.09
2        Coffee Shop  0.09
3             Bakery  0.09
4  Convenience Store  0.04


*****Business reply mail Processing Centre, South Central Letter Processing Plant Toronto*****
                  venue  freq
0  Gym / Fitness Center  0.06
1               Brewery  0.06
2            Skate Park  0.06
3            Restaurant  0.06
4           Pizza Place  0.06


*****CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport*****
                 venue  freq
0      Airport Service  0.17
1       Airport Lounge  0.11
2     Airport Terminal  0.11
3             Boutique  0.06
4  Rental Car Location  0.06


*****Central Bay St

## Let's put given results into a pandas dataframe


#### At first we created a function that sort the venues in descending order.

In [37]:
def return_most_common_venues(row, top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:top_venues]

#### Then we created a new dataframe and display the top 10 venues for each neighborhood.¶

In [38]:
top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...           Yoga Studio   
3   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville          Dessert Shop   
9                                    Davisville North                   Gym   
10                       Dufferin, Dovercourt Village              Pharmacy   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park         Jewelry Store   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
15                      High Park, The Junction South       Thai Restaurant   
16                     India Bazaar, The Beaches West        Sandwich Place   
17          Kensington Market, Chinatown, Grange Park                  Café   
18                                      Lawrence Park                  Park   
19                           Little Portugal, Trinity                   Bar   
20                        Moore Park, Summerhill East                   Gym   
21                  North Toronto West, Lawrence Park        Clothing Store   
22                             Parkdale, Roncesvalles        Breakfast Spot   
23        Queen's Park, Ontario Provincial Government           Coffee Shop   
24                          Regent Park, Harbourfront           Coffee Shop   
25                           Richmond, Adelaide, King           Coffee Shop   
26                                           Rosedale                  Park   
27                                           Roselawn        Ice Cream Shop   
28                                 Runnymede, Swansea           Coffee Shop   
29                                     St. James Town           Coffee Shop   
30                        St. James Town, Cabbagetown                  Café   
31                                     Stn A PO Boxes           Coffee Shop   
32                                    Studio District                  Café   
33  Summerhill West, Rathnelly, South Hill, Forest...           Coffee Shop   
34                The Annex, North Midtown, Yorkville        Sandwich Place   
35                                        The Beaches                 Trail   
36                       The Danforth West, Riverdale      Greek Restaurant   
37           Toronto Dominion Centre, Design Exchange           Coffee Shop   
38                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue      3rd Most Common Venue  \
0           Cocktail Bar                Cheese Shop   
1            Coffee Shop             Breakfast Spot   
2          Auto Workshop                       Park   
3         Airport Lounge           Airport Terminal   
4                   Café         Italian Restaurant   
5                   Café                       Park   
6       Sushi Restaurant        Japanese Restaurant   
7             Restaurant                       Café   
8         Sandwich Place           Sushi Restaurant   
9                  Hotel                       Park   
10                Bakery  Middle Eastern Restaurant   
11                  Café                        Gym   
12                 Trail           Sushi Restaurant   
13           Coffee Shop            Bubble Tea Shop   
14              Aquarium                       Café   


In [39]:
neighborhoods_venues_sorted.shape

(39, 11)

# Let's Cluster the Neighbourhoods

#### We use k-means algorithm to cluster the neighbourhood into clusters.

In [40]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

#### Let's create a new dataframe that includes assigned cluster as well as the top 10 venues for each neighborhood.

In [41]:
toronto_merged = toronto_DF

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_
toronto_merged.columns = toronto_merged.columns.astype(str)
toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels  
0  43.654260 -79.360636               1  
1  43.662301 -79.389494               1  
2  43.657162 -79.378937               1  
3  43.651494 -79.375418               1  
4  43.676357 -79.293031               1

In [42]:
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.676357 -79.293031               1                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1      Sushi Restaurant        Discount Store        Sandwich Place   
2           Coffee Shop       Bubble Tea Shop   Japanese Restaurant   
3                  Café          Cocktail Bar             Gastropub   
4     Health Food Store                   Pub           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0        Breakfast Spot                  Café                 Restaurant   
1                  Park           Music Venue         Mexican Restaurant   
2    Italian Restaurant                  Café  Middle Eastern Restaurant   
3   American Restaurant            Restaurant                      Hotel   
4                 Diner        Discount Store        Distribution Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Theater        Farmers Market         Chocolate Shop  
1    Italian Restaurant            Hobby Shop    Fried Chicken Joint  
2        Cosmetics Shop                 Diner               Tea Room  
3    Italian Restaurant        Lingerie Store       Department Store  
4               Dog Run      Doner Restaurant             Donut Shop

# Final results - Clusters visualization


In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### END OF PART 3